nova/src/circuits/supernova/augmented.rs

```diff
@@ -0,0 +1,566 @@
+    }
+
+    fn select<T: CondSelectGadget<F>>(&self, vals: &[T]) -> Result<T, SynthesisError> {
+        let bits = &self.bits;
+        assert_eq!(vals.len(), bits.len());
+
+        // skipping first bit is correct since all bits are constrained.
+        let mut val = vals.first().cloned().expect("must be non-empty");
+        for (v, bit) in vals.iter().zip(bits).skip(1) {
+            val = bit.select(v, &val)?;
+        }
+        Ok(val)
+    }
+
+    fn insert<T: CondSelectGadget<F>>(
+        &self,
+        vals: &[T],
+        ival: &T,
+    ) -> Result<Vec<T>, SynthesisError> {
+        let bits = &self.bits;
+        assert_eq!(vals.len(), bits.len());
+
+        vals.iter()
+            .zip(bits)
+            .map(|(val, bit)| bit.select(ival, val))
+            .collect::<Result<_, _>>()
+    }
+}
+
+impl<G1, G2, C1, C2, RO, SC> SuperNovaConstraintSynthesizer<G1::ScalarField>
+    for SuperNovaAugmentedCircuit<'_, G1, G2, C1, C2, RO, SC>
+where
+    G1: SWCurveConfig,
+    G2: SWCurveConfig<BaseField = G1::ScalarField, ScalarField = G1::BaseField>,
+    G1::BaseField: PrimeField + Absorb,
+    G2::BaseField: PrimeField + Absorb,
+    C1: CommitmentScheme<Projective<G1>>,
+    C2: CommitmentScheme<Projective<G2>>,
+    RO: SpongeWithGadget<G1::ScalarField>,
+    RO::Var: CryptographicSpongeVar<G1::ScalarField, RO, Parameters = RO::Config>,
+    SC: NonUniformCircuit<G1::ScalarField>,
+{
+    fn generate_constraints(
+        self,
+        cs: ConstraintSystemRef<G1::ScalarField>,
+    ) -> Result<(FpVar<G1::ScalarField>, Vec<FpVar<G1::ScalarField>>), SynthesisError> {
+        let input = SuperNovaAugmentedCircuitInputVar::<G1, G2, C1, C2, RO>::new_witness(
+            cs.clone(),
+            || Ok(&self.input),
+        )?;
+        let num_augmented_circuits = input.U.len();
+        assert_eq!(input.U_secondary.len(), num_augmented_circuits);
+
+        let is_base_case = input.i.is_zero()?;
+        let should_enforce = is_base_case.not();
+
+        let U_base = primary::RelaxedR1CSInstanceVar::<G1, C1>::new_constant(
+            cs.clone(),
+            RelaxedR1CSInstance {
+                commitment_W: Projective::zero().into(),
+                commitment_E: Projective::zero().into(),
+                X: vec![G1::ScalarField::ZERO; AUGMENTED_CIRCUIT_NUM_IO],
+            },
+        )?;
+        let U_secondary_base = secondary::RelaxedR1CSInstanceVar::<G2, C2>::new_constant(
+            cs.clone(),
+            RelaxedR1CSInstance {
+                commitment_W: Projective::zero().into(),
+                commitment_E: Projective::zero().into(),
+                X: vec![G2::ScalarField::ZERO; SecondaryCircuit::<G1>::NUM_IO],
+            },
+        )?;
+
+        for (z_0, z_i) in input.z_0.iter().zip(&input.z_i) {
+            z_0.conditional_enforce_equal(z_i, &is_base_case)?;
+        }
+        input
+            .pc
+            .conditional_enforce_equal(&input.pc_prev, &is_base_case)?;
+        input
```

Comment:
this constraint is unnecessary. `pc`, which should be computed as _output_ of this circuit, doesn't need to be tied in-circuit to the index of the circuit. this circuit will enforce that `pc_prev` is part of the data hashed into the output of `u`, and fold `u` into `U[pc_prev]`. this then ensures that the index of the circuit for `u` is `pc_prev` (since the final decider will test `U[j]` against the circuit with index `j`)

Reply:
This is a 1 constraint safe guard, your scenario leads to invalid proof as well, but I don't see no harm in it.